### PR TITLE
[BugFix] Distinguish join's on-predicates and where-predicates for non inner/cross join

### PR DESCRIPTION
**Pull request by LiShuMing**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Problem Summary(Required) ：
- Distinguish join's on-predicates and where-predicates for non inner/cross join;
- For non-inner/cross join, all on-predicates must be rewritten, otherwise may cause correct bugs.

```
  // If join is not cross/inner join, MV Rewrite must rewrite, otherwise may cause bad results.
    // eg.
    // mv: select * from customer left outer join orders on c_custkey = o_custkey;
    //
    //query（cannot rewrite）:
    //select count(1) from customer left outer join orders
    //  on c_custkey = o_custkey and o_comment not like '%special%requests%';
    //
    //query（can rewrite）:
    //select count(1)
    //          from customer left outer join orders on c_custkey = o_custkey
    //          where o_comment not like '%special%requests%';
```

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java [hunk 1]
```diff
@@ -382,6 +382,33 @@ public static List<ScalarOperator> getAllPredicates(OptExpression root) {
         return predicates;
     }
 
+    // If join is not cross/inner join, MV Rewrite must rewrite, otherwise may cause bad results.
+    // eg.
+    // mv: select * from customer left outer join orders on c_custkey = o_custkey;
+    //
+    //query（cannot rewrite）:
+    //select count(1) from customer left outer join orders
+    //  on c_custkey = o_custkey and o_comment not like '%special%requests%';
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java:391]: why would user write such SQL

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java:391]: Yep.... We cannot decide it. But for non-inner join, we should take care of on-predicates or where-predicates.

**Review by murphyatwork — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/RangeSimplifier.java [hunk 2]
```diff
@@ -43,50 +42,12 @@ public RangeSimplifier(List<ScalarOperator> srcPredicates) {
...
-                }
-                Range<ConstantOperator> columnRange = targetColumnIdToRange.get(targetColumn.getId());
-                Range<ConstantOperator> range = range(srcBinary.getBinaryType(), targetConstant);
-                columnRange = columnRange.intersection(range);
-                if (columnRange.isEmpty()) {
-                    return null;
-                }
-                targetColumnIdToRange.put(targetColumn.getId(), columnRange);
-            }
+            Map<Integer, Range<ConstantOperator>> srcColumnIdToRange = extractColumnIdRangeMapping(srcPredicates);
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/RangeSimplifier.java:45]: if (srcColumnIdToRange == null) {
    return null;
}

if (targetColumnIdToRange == null) {
    return null;
}

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 3]
```diff
@@ -1268,18 +1273,39 @@ private Map<Integer, Map<String, ColumnRefOperator>> getRelationIdToColumns(Colu
         return result;
     }
 
+    private boolean isCompensationComplete(ScalarOperator rewritten,
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:1276]: isCompensationComplete is not a good name.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 4]
```diff
@@ -1268,18 +1273,39 @@ private Map<Integer, Map<String, ColumnRefOperator>> getRelationIdToColumns(Colu
...
         // 1. equality join subsumption test
         EquivalenceClasses sourceEquivalenceClasses = isQueryAgainstView ?
                 rewriteContext.getQueryEquivalenceClasses() : rewriteContext.getQueryBasedViewEquivalenceClasses();
         EquivalenceClasses targetEquivalenceClasses = isQueryAgainstView ?
                 rewriteContext.getQueryBasedViewEquivalenceClasses() : rewriteContext.getQueryEquivalenceClasses();
         final ScalarOperator compensationEqualPredicate =
                 getCompensationEqualPredicate(sourceEquivalenceClasses, targetEquivalenceClasses);
-        if (compensationEqualPredicate == null) {
-            // means source equal predicates cannot be rewritten by target
+        if (!isCompensationComplete(compensationEqualPredicate, joinOnPredicateColumnRefSet)) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:1308]: if query is:
select count(1)
    //          from customer left outer join orders on c_custkey = o_custkey
    //          where  c_custkey = 1
can this query pass this check?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/RangeSimplifier.java (hunk 2 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/RangeSimplifier.java:45]: done.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java (hunk 3 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java]: fix it.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java (hunk 4 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java]: Add a unit test.

**Review by kangkaisen — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**